File: python/solution_02.py
from utils import get_input_lines, read_input


def get_dimensions(present: str) -> list[int]:
    return [int(d) for d in present.split("x")]


def get_areas(present_dimensions: list[int]) -> list[int]:
    return [
        (present_dimensions[0] * present_dimensions[1]),
        (present_dimensions[1] * present_dimensions[2]),
        (present_dimensions[0] * present_dimensions[2]),
    ]


def get_perimeters(present_dimensions: list[int]) -> list[int]:
    return [
        (2 * present_dimensions[0]) + (2 * present_dimensions[1]),
        (2 * present_dimensions[1]) + (2 * present_dimensions[2]),
        (2 * present_dimensions[0]) + (2 * present_dimensions[2]),
    ]


def get_volume(present_dimensions: list[int]) -> int:
    return present_dimensions[0] * present_dimensions[1] * present_dimensions[2]


def part_1(input: list[str]) -> int:
    paper_needed: int = 0
    for present in input:
        if present != "":
            present_areas = get_areas(get_dimensions(present))
            paper_needed += (
                (2 * present_areas[0])
                + (2 * present_areas[1])
                + (2 * present_areas[2])
                + min(present_areas)
            )
    return paper_needed


def part_2(input: str) -> int:
    ribbon_needed: int = 0
    for present in input:
        if present != "":
            present_dimensions = get_dimensions(present)
            present_perimeters = get_perimeters(present_dimensions)
            ribbon_needed += min(present_perimeters) + get_volume(present_dimensions)
    return ribbon_needed


def solve() -> None:
    input = get_input_lines(read_input("02"))
    return part_1(input), part_2(input)


if __name__ == "__main__":
    paper, ribbon = solve()
    print(f"Wrapping Paper Needed: {paper}")
    print(f"Ribbon Needed: {ribbon}")

File: rust/src/solutions/solution_02.rs
fn get_dimensions(present: &str) -> [i32; 3] {
    let mut present_dimensions: [i32; 3] = [0; 3];
    let mut i: usize = 0;
    for dimension in present.split("x") {
        present_dimensions[i] = dimension.parse::<i32>().unwrap();
        i += 1;
    }
    return present_dimensions;
}

fn get_areas(present_dimensions: [i32; 3]) -> [i32; 3] {
    let mut present_areas: [i32; 3] = [0; 3];
    present_areas[0] = present_dimensions[0] * present_dimensions[1];
    present_areas[1] = present_dimensions[1] * present_dimensions[2];
    present_areas[2] = present_dimensions[0] * present_dimensions[2];
    return present_areas;
}

fn get_perimeters(present_dimensions: [i32; 3]) -> [i32; 3] {
    let mut present_perimeters: [i32; 3] = [0; 3];
    present_perimeters[0] = (2 * present_dimensions[0]) + (2 * present_dimensions[1]);
    present_perimeters[1] = (2 * present_dimensions[1]) + (2 * present_dimensions[2]);
    present_perimeters[2] = (2 * present_dimensions[0]) + (2 * present_dimensions[2]);
    return present_perimeters;
}

fn get_volume(present_dimensions: [i32; 3]) -> i32 {
    return present_dimensions[0] * present_dimensions[1] * present_dimensions[2];
}

fn part_1(lines: &Vec<&str>) -> i32 {
    let mut total_needed: i32 = 0;
    for present in lines {
        if !present.is_empty() {
            let present_dimensions: [i32; 3] = get_dimensions(present);
            let present_areas: [i32; 3] = get_areas(present_dimensions);
            total_needed += (2 * present_areas[0])
                + (2 * present_areas[1])
                + (2 * present_areas[2])
                + present_areas.iter().min().unwrap();
        }
    }
    return total_needed;
}

fn part_2(lines: &Vec<&str>) -> i32 {
    let mut total_needed: i32 = 0;
    for present in lines {
        if !present.is_empty() {
            let present_dimensions: [i32; 3] = get_dimensions(present);
            let present_perimeters: [i32; 3] = get_perimeters(present_dimensions);
            total_needed += present_perimeters.iter().min().unwrap()
                + get_volume(present_dimensions);
        }
    }
    return total_needed;
}

pub fn solve(input: Vec<&str>) -> (i32, i32) {
    return (part_1(input.as_ref()), part_2(input.as_ref()));
}
